datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,344 @@ fn optimize_join(
-                // predicates reference the shared join columns are pushed to both sides.
-                // This logic should also been applied to conditions in JOIN ON clause
-                let join_side_filters = state
-                    .filters
+            }
+            LogicalPlan::Projection(projection) => {
+                // A projection is filter-commutable, but re-writes all predicate expressions
+                // collect projection.
+                let replace_map = projection
+                    .schema
+                    .fields()
                     .iter()
-                    .chain(on_filters.iter())
-                    .filter_map(|(predicate, columns)| {
-                        let mut join_cols_to_replace = HashMap::new();
-                        for col in columns.iter() {
-                            for (l, r) in on {
-                                if col == l {
-                                    join_cols_to_replace.insert(col, r);
-                                    break;
-                                } else if col == r {
-                                    join_cols_to_replace.insert(col, l);
-                                    break;
-                                }
-                            }
-                        }
-
-                        if join_cols_to_replace.is_empty() {
-                            return None;
-                        }
-
-                        let join_side_predicate =
-                            match replace_col(predicate.clone(), &join_cols_to_replace) {
-                                Ok(p) => p,
-                                Err(e) => {
-                                    return Some(Err(e));
-                                }
-                            };
-
-                        let join_side_columns = columns
-                            .clone()
-                            .into_iter()
-                            // replace keys in join_cols_to_replace with values in resulting column
-                            // set
-                            .filter(|c| !join_cols_to_replace.contains_key(c))
-                            .chain(join_cols_to_replace.values().map(|v| (*v).clone()))
-                            .collect();
-
-                        Some(Ok((join_side_predicate, join_side_columns)))
+                    .enumerate()
+                    .flat_map(|(i, field)| {
+                        // strip alias, as they should not be part of filters
+                        let expr = match &projection.expr[i] {
+                            Expr::Alias(expr, _) => expr.as_ref().clone(),
+                            expr => expr.clone(),
+                        };
+
+                        // Convert both qualified and unqualified fields
+                        [
+                            (field.name().clone(), expr.clone()),
+                            (field.qualified_name(), expr),
+                        ]
```

Comment:
I understand this is coming from the existing logic, but could it encounter conflict unqualified field names here?

Reply:
Yes, original code here exist bug.

Reply:
has resolved this bug, and add UT, thanks @mingmwang !